_posts/2023-02-01-user-defined-operators.md

```diff
@@ -0,0 +1,173 @@
+
+The following example shows off the main features of `using for` in combination wth UDVTs:
+
+### Introducing User-Defined Operators
+The new feature adds another form of the `using for` syntax, where instead of attaching a function,
+it is assigned an operator symbol.
+The function is then invoked whenever the operator is used on the values of the type:
+
+```solidity
+type Int is int;
+using {add as +} for Int;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a + b;
+}
+```
+
+Note that defining an operator for a type does not automatically make it available as an attached function.
+E.g. `using {add as +} for Int` does not imply `using {add} for Int`.
+
+#### Which Operators Can Be User-Defined?
+| Category   | Operators                        |
+|------------|----------------------------------|
+| Bitwise    | `&`, `\|`, `^`,  `~`             |
+| Arithmetic | `+`, `-`, `*`, `/`, `%`          |
+| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |
+
+Of the operators listed above `!` and `~` are unary, `-` can be both unary and binary and the
+rest is always binary.
+Note that Solidity does not have binary `+`.
+
+Several operators have been excluded from the initial implementation for the sake of simplicity.
+We are considering enabling them in one of the future releases if the design issues can be
+resolved satisfactorily:
+1. `&&`, `||` and `!` are not necessary for most UDVTs because user-defined comparison operators
+    must return `bool`, for which the built-in boolean operators are sufficient.
+
+    In case of `&&` and `||`  an additional complication is their special short-circuiting behavior.
+1. `<<`, `>>` and `**` have non-uniform arguments.
+    For built-in types the right-hand-side is always an unsigned type.
+1. `=` is one of the few operators that already work with UDVTs.
+    The new feature is not meant to allow redefining existing operators.
+1. Assignment and increment/decrement operators in general modify their argument and return a reference.
+
+    `++` and `--` also have separate prefix and postfix variants with different semantics.
+
+#### Which Types Can Have User-Defined Operators?
+Currently operators can only be defined for user-defined value types.
+We are also considering enabling them for structs and enums in future releases.
+There are no plans to allow defining operators for built-in value types, especially in cases where
+an operator is already provided by the compiler.
+
+Note that the type wildcard (`*`) is not allowed in `using for` when defining operators, which
+is in line with how it already worked when used attach individual functions.
+
+Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
+must always be invoked with exact types.
+
```

Comment:
```suggestion
#### Which Types Can Have User-Defined Operators?
Currently, operators can only be defined for user-defined value types.
We are also considering enabling them for structs and enums in future releases.
There are no plans, for now, to allow defining operators for built-in value types, especially in cases where
an operator is already provided by the compiler.

Note that the type wildcard (`*`) is not allowed in `using for` when defining operators, which
is in line with how it already worked when used to attach individual functions.

Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
must always be invoked with exact types.

```